fhevm-engine/transaction-sender/src/ops/add_ciphertext.rs

```
@@ -1,33 +1,157 @@
+use alloy::{
+    network::{Ethereum, TransactionBuilder},
+    primitives::{Address, FixedBytes, U256},
+    providers::Provider,
+    rpc::types::TransactionRequest,
+    sol,
+};
 use async_trait::async_trait;
+use fhevm_engine_common::{tenant_keys::query_tenant_info, utils::to_hex};
 use sqlx::{Pool, Postgres};
+use tokio::task::JoinSet;
+use tracing::{error, info};
 
 sol!(
     #[sol(rpc)]
-    CiphertextStorage,
-    "artifacts/CiphertextStorage.sol/CiphertextStorage.json"
+    CiphertextManager,
+    "artifacts/CiphertextManager.sol/CiphertextManager.json"
 );
 
+macro_rules! try_into_array {
+    ($vec:expr, $size:expr) => {{
+        if $vec.len() != $size {
+            panic!("invalid len, expected {} but got {}", $size, $vec.len());
+        }
+        let arr: [u8; $size] = $vec.try_into().expect("Failed to convert Vec to array");
+        arr
+    }};
+}
+
 #[derive(Clone)]
 pub struct AddCiphertextOperation<P: Provider<Ethereum> + Clone + 'static> {
-    ciphertext_storage_address: Address,
+    ciphertext_manager_address: Address,
     provider: P,
     conf: crate::ConfigSettings,
+    gas: Option<u64>,
+}
+
+impl<P: Provider<Ethereum> + Clone + 'static> AddCiphertextOperation<P> {
+    async fn send_transaction(
+        db_pool: Pool<Postgres>,
+        provider: P,
+        handle: Vec<u8>,
+        txn_request: impl Into<TransactionRequest>,
+    ) -> anyhow::Result<()> {
+        let h = to_hex(&handle);
+
+        info!("Processing transaction, handle: {}", h);
+
+        let txn_req = txn_request.into();
+        let transaction = match provider.send_transaction(txn_req.clone()).await {
+            Ok(txn) => txn,
+            Err(e) => {
+                error!(
+                    "Transaction {:?} sending failed with error: {}, handle: {}",
+                    txn_req, e, h
+                );
+
+                Self::increment_db_retry(&db_pool, handle, &e.to_string()).await?;
+                return Ok(());
```

Shouldn't we return the error here?

---

Good catch.

---

Fixed